importer/src/path_canonicalizer.rs

```diff
@@ -0,0 +1,378 @@
+                    match symbol {
+                        Symbol::Machine(m) => Some(Symbol::Machine(canonicalize_machine(
+                            absolute_path.clone(),
+                            m,
+                            paths,
+                        ))),
+                        Symbol::Import(_) => None,
+                        Symbol::Module(m) => match m {
+                            Module::External(_) => unreachable!(),
+                            Module::Local(m) => {
+                                Some(Symbol::Module(Module::Local(canonicalize_paths_in_module(
+                                    absolute_path.clone().join(name.clone()),
+                                    m,
+                                    paths,
+                                ))))
+                            }
+                        },
+                    }
+                    .map(|symbol| SymbolDefinition { name, symbol }.into())
+                }
+            })
+            .collect(),
+    }
+}
+
+fn canonicalize_machine<T: FieldElement>(
+    absolute_path: AbsoluteSymbolPath,
+    mut m: Machine<T>,
+    symbols: &PathMap,
+) -> Machine<T> {
+    for s in &mut m.statements {
+        if let MachineStatement::Submachine(_, path, _) = s {
+            *path = symbols
+                .get(&(absolute_path.clone(), std::mem::take(path)))
+                .cloned()
+                .unwrap()
+                .into();
+        }
+    }
+
+    m
+}
+
+/// Answers the question: When at absolute path `p`, refering to relative path `r`, what is the absolute path to the canonical symbol imported?
+pub type PathMap = BTreeMap<(AbsoluteSymbolPath, SymbolPath), AbsoluteSymbolPath>;
+
+#[derive(PartialEq, Debug)]
+pub struct State<'a, T> {
+    /// the root module of this program
+    root: &'a ASMModule<T>,
+    /// the paths to replace
+    pub paths: PathMap,
+}
+
+/// Checks a relative path in the context of an absolute path
+///
+/// # Panics
+///
+/// Panics if we encounter an external module which wasn't turned into a local one
+///
+/// # Errors
+///
+/// This function will return an error if the relative path does not resolve to anything
+fn check_path<T: Clone>(
+    // the location at which the import is made
+    location: AbsoluteSymbolPath,
+    // the imported path, relative to the location
+    imported: SymbolPath,
+    // the current state
+    state: State<'_, T>,
+) -> Result<(State<'_, T>, AbsoluteSymbolPath, Symbol<T>), String> {
+    let root = state.root.clone();
+    // walk down the tree of modules
+    location
+        .clone()
+        .join(imported.clone())
+        .parts
+        .iter()
+        .try_fold(
+            (
+                state,
+                AbsoluteSymbolPath::default(),
+                Symbol::Module(Module::Local(root)),
+            ),
+            |(state, mut location, symbol), member| {
+                match symbol {
+                    // machines do not expose symbols
```

Comment:
So I cannot even access a symbol within the same machine?

I think we should abstract that to some more generic terms like namespaces, scopes, etc.